### PR TITLE
Support CenterFormer in `projects`

### PR DESCRIPTION
**Pull request by JingweiZhang12**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by JingweiZhang12** [comment]
The model and log will be uploaded soon.

File: projects/centerformer/centerformer/bbox_ops.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import torch
+from mmcv.utils import ext_loader
+
+ext_module = ext_loader.load_ext('_ext', ['iou3d_nms3d_forward'])
+
+
+def nms_iou3d(boxes, scores, thresh, pre_maxsize=None, post_max_size=None):
+    """NMS function GPU implementation (using IoU3D) The difference of this
```
Comment (ZCMax) [inline projects/centerformer/centerformer/bbox_ops.py:8]: The difference between this implementation and nms3d in MMCV is that we add `pre_maxsize` and `post_max_size` before and after NMS respectively.

File: projects/centerformer/centerformer/centerformer.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        return hasattr(self,
+                       'voxel_encoder') and self.voxel_encoder is not None
+
+    @property
+    def with_middle_encoder(self):
+        """bool: Whether the detector has a middle encoder."""
+        return hasattr(self,
+                       'middle_encoder') and self.middle_encoder is not None
+
+    def _forward(self):
```
Comment (ZCMax) [inline projects/centerformer/centerformer/centerformer.py:108]: some of the methods should already be implemented in `BaseDetector`

File: projects/centerformer/centerformer/centerformer.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        pass
+
+    def extract_feat(self, batch_inputs_dict: dict,
+                     batch_input_metas: List[dict]) -> tuple:
```
Comment (ZCMax) [inline projects/centerformer/centerformer/centerformer.py:112]: It seems `batch_input_metas` is not used in `extract_feat`

Comment (ZCMax) [reply projects/centerformer/centerformer/centerformer.py:112]: please check whether we need this parameter in `loss` and `predict`

File: projects/centerformer/README.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
+# CenterFormer: Center-based Transformer for 3D Object Detection
```
Comment (ZCMax) [inline projects/centerformer/README.md:1]: I recommend to rename this project to `CenterFormer`, the path should be `projects/CenterFormer/centerformer/...`

File: projects/centerformer/centerformer/centerformer.py (hunk 3 above)
Comment (JingweiZhang12) [reply projects/centerformer/centerformer/centerformer.py:112]: It's not used but `batch_input_metas` or `batch_data_samples` are the default arguments in some high-level function. I suggest keeping it.

Comment (ZCMax) [reply projects/centerformer/centerformer/centerformer.py]: OK I got it

**Review by ZCMax — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]
